Add TaskState to worker.py

gforsyth opened the pull request:
This is some initial progress towards #4097 -- dependents, dependencies and dep_state are still being tracked in a single dict on the Worker class but those come next.

Still lots to do on this, so no need for anyone to review just yet.

mrocklin commented:
Hooray for PRs with incremental progress!

FWIW I really like the approach of starting with just the task-wise state.  That seems sanity-preserving.

mrocklin commented:
Checking in here @gforsyth .  How are things going?  Have you run into tricky bits that I can help with?

gforsyth commented:
> How are things going?

Definitely hit some tricky bits, but there's something I thought of I want to try.  
Mostly working, just some bookkeeping issues w.r.t. updating `data_needed`, `waiting_for_data`, etc. in a sane way.

I'll ping if the next few hours don't resolve things.

mrocklin commented:
It's cool seeing all of the activity here :)  

How are things going for you @gforsyth ?

gforsyth commented:
> How are things going for you @gforsyth ?

I think I'm closing in on a decent first cut of this.  I have a few more test failures I'm trying to trace out.

mrocklin commented on distributed/worker.py:
```diff
@@ -1602,6 +1606,25 @@ def transition_waiting_ready(self, ts):
 
+    def transition_ready_waiting(self, ts, worker):
+        # TODO: do we actually want this?  or is this a bandaid for
+        # bad bookkeeping elsewhere?
```
I'm curious.  What caused the need for this?

gforsyth replied:
I think this is actually a result of some bad bookkeeping, but I stuck it in there to both give me a place to set a breakpoint, but also to see where else things broke if there wasn't a `KeyError` in `self._transitions`.  It should not be required once I've sorted that out.

mrocklin commented on distributed/worker.py:
```diff
@@ -1485,7 +1486,10 @@ def add_task(
             raise
 
     def transition(self, key, finish, **kwargs):
-        ts = self.tasks[key]
```
I'm curious, what cause the need for this?

mrocklin reviewed (commented):
Some minor comments. I realized that I had some other comments queued up here from a previous review that I neglected to send.  It looks like a few of those are obsolete now.

gforsyth commented on distributed/worker.py:
```diff
@@ -1435,34 +1436,34 @@ def add_task(
             #    }
             # dep -> dependency (parent)
-            for parent, workers in who_has.items():
+            for dependency, workers in who_has.items():
```
ahh, yeah, this was a previous bit of disambiguation -- elsewhere I've stuck to being a little more verbose and saying `dependent` and `dependency`, I'll also make that change here.

gforsyth commented on distributed/worker.py:
```diff
@@ -1338,7 +1339,7 @@ async def get_data(
     def update_data(self, comm=None, data=None, report=True, serializers=None):
         for key, value in data.items():
-            if key in self.tasks:
+            if getattr(self.tasks.get(key), "state", None) is not None:
```
This shouldn't happen anymore.  It was possible for a task to be instantiated with a default state of `None`, but I think setting it to `"waiting"` at instantiation nixes this issue.

mrocklin commented:
Ooh, nice green check marks :heavy_check_mark:

mrocklin commented:
@gforsyth are we ready to remove the WIP label from this PR?  Would it be reasonable to merge it?

gforsyth commented:
> @gforsyth are we ready to remove the WIP label from this PR?  Would it be reasonable to merge it?

I can remove the WIP. I think it's ready for an initial review. There are comments to clean up and some docstrings to expand a bit. 

I think there's more that _could_ be done, but probably not much more that _needs_ to be done. And given that I'm going to try out speculative task assignment it might make more sense to merge this in closer to as-is (pending review) than to get it "perfect" before tearing it apart again.

mrocklin commented on distributed/worker.py:
```diff
@@ -1545,18 +1539,22 @@ def transition_dep_flight_waiting(self, dep, worker=None, remove=True):
+            del self.in_flight_tasks[ts.key]
+            if ts.dependents:
+                self.put_key_in_memory(ts, value)
+                assert ts.key in self.data
```
If the value is large enough to be placed into slow storage immediately (unlikely) then this might not be free.  We might want to drop it.

mrocklin commented on distributed/worker.py:
```diff
@@ -1663,52 +1634,52 @@ def transition_ready_executing(self, key):
+    def transition_executing_done(self, ts, value=no_value, report=True):
         try:
             if self.validate:
-                assert key in self.executing or key in self.long_running
```
Is it possible to reinstate this check?

gforsyth replied:
Yep!

mrocklin commented on distributed/worker.py:
```diff
@@ -1663,52 +1634,52 @@ def transition_ready_executing(self, key):
+                # Don't release the dependency keys, but do remove them from `dependents`
+                for dependency in ts.dependencies:
+                    dependency.dependents.discard(ts)
+                ts.dependencies = set()
```
Tiny nit, but maybe `.clear()`?

If this was in the the scheduler I would recommend this due to what I would expect to be a small performance gain.  It probably doesn't matter as much here though.

gforsyth replied:
Sounds good.

mrocklin commented on distributed/worker.py:
```diff
@@ -1792,31 +1767,32 @@ def ensure_communicating(self):
+                deps = ts.dependencies
                 if self.validate:
-                    assert all(dep in self.dep_state for dep in deps)
+                    assert all(hasattr(dep, "state") for dep in deps)
```
I would think that TaskState objects always have a `.state` attribute

gforsyth replied:
yeah, I think this was from an earlier debugging pass, cleaning it up now.

mrocklin commented on distributed/worker.py:
```diff
@@ -1792,31 +1767,32 @@ def ensure_communicating(self):
-                deps = self.dependencies[key]
+                deps = ts.dependencies
                 if self.validate:
-                    assert all(dep in self.dep_state for dep in deps)
```
Should this be `dep in self.tasks` now?

gforsyth replied:
Yes, I think so.  Thanks

mrocklin commented on distributed/worker.py:
```diff
@@ -1843,18 +1819,21 @@ def ensure_communicating(self):
                     self.in_flight_workers[worker] = to_gather
                     for d in to_gather:
-                        self.transition_dep(d, "flight", worker=worker)
+                        self.transition(self.tasks.get(d), "flight", worker=worker)
```
Is it safe to do `self.tasks[d]` here, or should we be concerned about `d not in self.tasks` as a possibility?

gforsyth replied:
I think it should be safe to do `self.tasks[d]` here, since the dependencies from `dep` should be created at this point.  I think I was worried about something dying in flight but that doesn't apply here.

mrocklin commented on distributed/worker.py:
```diff
@@ -1884,73 +1863,70 @@ def send_task_state_to_scheduler(self, key):
         while L:
             d = L.popleft()
-            if self.dep_state.get(d) != "waiting":
+            if getattr(self.tasks.get(d), "state", None) != "waiting":
```
Do we still need this `getattr` call or can we assume this?

Oh, or is this for if `d not in self.tasks`?

gforsyth replied:
> Oh, or is this for if `d not in self.tasks`?

Yeah, I should add a comment about that

gforsyth replied:
Or just make it more explicit in the first place

mrocklin commented on distributed/worker.py:
```diff
@@ -1960,6 +1936,12 @@ def select_keys_for_gather(self, worker, dep):
 
     async def gather_dep(self, worker, dep, deps, total_nbytes, cause=None):
+        """
+        worker: str
+        dep: TaskState object
+        deps: list of keys of dependencies to gather
```
I support docstrings generally, but when we add them we should probably adhere to numpdoc standard.  

I know that you know this stuff better than 99.999999% of the population (probably an underestimate) but I'll do what I always do, and copy in this doc: https://docs.dask.org/en/latest/develop.html#docstrings

gforsyth replied:
Yep, these were notes to myself that I forgot to remove, but I'll make it a proper docstring instead (notes for my future self)

mrocklin commented on distributed/worker.py:
```diff
@@ -1970,9 +1952,11 @@ async def gather_dep(self, worker, dep, deps, total_nbytes, cause=None):
-                deps = tuple(dep for dep in deps if self.dep_state.get(dep) == "flight")
+                deps_ts = [self.tasks[key] for key in deps]
+                deps_ts = tuple(ts for ts in deps_ts if ts.state == "flight")
+                deps = [d.key for d in deps_ts]
```
This really makes you want variable scope in list comprehensions :)

gforsyth replied:
YES

mrocklin commented on distributed/worker.py:
```diff
@@ -2037,9 +2022,9 @@ async def gather_dep(self, worker, dep, deps, total_nbytes, cause=None):
-                        del self.who_has[d]
+                    self.tasks[d].who_has.remove(worker)
+                # If the worker has died, `return` here so we don't hit `finally`
+                return
```
??  What happened here?

gforsyth replied:
I genuinely don't remember why I thought this was necessary but tests pass locally when I remove it.

mrocklin commented on distributed/worker.py:
```diff
@@ -2465,28 +2396,37 @@ def ensure_computing(self):
+                    # It is possible for tasks to be released while still remaining on `ready`
+                    # The scheduler might have re-routed to a new worker and told this worker
+                    # to release.  If the task has "disappeared" just continue through the heap
+                    continue
```
Thanks for taking the time to encode this explanation

mrocklin commented on distributed/worker.py:
```diff
@@ -2843,86 +2790,46 @@ def _notify_plugins(self, method_name, *args, **kwargs):
-            elif state == "memory":
-                self.validate_dep_memory(dep)
-            else:
-                raise ValueError("Unknown dependent state", state)
```
Just checking, can you verify that these checks were moved to the other validate methods?

Most of our testing confidence comes from running these methods while operating Dask in different situations.  It's very important to us that we not lose validation here.

gforsyth replied:
Thanks for the call out.  Deps are gone now, but all of these checks (except for `flight`) are covered in `validate_task_state`.  I'm adding an additional check for `flight` now, which should then cover all of these bases.

mrocklin reviewed (commented):
A bunch of tiny comments!

mrocklin commented:
Also, kudos on adding capability while dropping line-count :)

mrocklin commented:
(unless that line count was due to dropping validation functions, in which case you will be hunted down by future generations of Dask maintainers)

gforsyth commented:
Thank you for the many small notes, @mrocklin.  It is my hope and belief that the line-count drop is mostly due to combining various validation functions, rather than removing them.  :crossed_fingers: that I am not the subject of the most dangerous game.

I think I've addressed all of your comments, but I'll take another cleaning pass now, too.

mrocklin commented on distributed/worker.py:
```diff
@@ -2521,58 +2477,62 @@ async def execute(self, key, report=False):
                 )
+                # TODO: it seems insane that this needs to be here
+                # but it is occasionally cleared by something in executor_submit
+                self.executing.add(ts.key)
```
Why does this need to be here?  What happens if it is removed?

Grepping through the codebase I see three paths for this:

1.  The task has been released during execution.  Maybe the user no longer needs the result
2.  The task is long-running, such as is called with `worker_client` or `secede`
3.  The same task is run twice (??)

In some of these situations I would expect the following `ts.state not in ("executing", "long-running")` check to clean things up.

gforsyth replied:
Hmm, 
I had put this in because there were tests in `test_failed_worker.py` that would throw assertion errors (but not fail the test in question) but I've just removed it and can no longer get those errors to show up, so it may have been due to a problem elsewhere that is now fixed.

jrbourbeau commented on distributed/worker.py:
```diff
@@ -2253,70 +2241,24 @@ def release_key(self, key, cause=None, reason=None, report=True):
                 pdb.set_trace()
             raise
 
-    def release_dep(self, dep, report=False):
```
It looks like we've replaced `release_dep` calls with `release_keys`. What was the reasoning for this? It'll impact worker plugins

gforsyth replied:
There's no longer a distinction between a `dep` and a `Task` which is why I removed it.  But if it's going to impact the plugins I can restore it and have it wrap `release_keys` (or we can update the plugins?)

mrocklin replied:
> It'll impact worker plugins

Do we know of any worker plugins that trigger on `release_dep`?

Also, If we're jumping from version 2 to 2020 in the next release then we can also break some things.  Breaking `WorkerPlugin.release_dep` sounds ok to me.

jrbourbeau replied:
> Do we know of any worker plugins that trigger on release_dep?

None come to mind, but there may be user-defined plugins that use `release_dep`

> Breaking WorkerPlugin.release_dep sounds ok to me

Yeah, I think we're treating all the HLG changes + the bump to calver as a major version change, so removing `release_dep` should be okay

jrbourbeau reviewed (commented):
Apologies for the delayed review @gforsyth. Thanks for all your careful work here!

gforsyth commented on distributed/worker.py:
```diff
@@ -706,7 +725,7 @@ def __init__(
 
     def __repr__(self):
-        return "<%s: %r, %s, %s, stored: %d, running: %d/%d, ready: %d, comm: %d, waiting: %d>" % (
+        return "<%s: %r, %s, %s, stored: %d, running: %d/%d, ready: %d, comm: %d>" % (
```
Restoring that now, I had removed `waiting` entirely from the `Worker` class but restored it so we don't have to do a loop over all pending tasks when we update this

gforsyth commented on distributed/worker.py:
```diff
@@ -91,6 +91,73 @@
+        The exception caused by running a task if it erred
+    * **traceback**: ``str``
+        The exception caused by running a task if it erred
+    * **type_**: ``type``
```
Yeah, I didn't want to use the name `type`, even as an attribute.  Happy to change this if it reads weird.

jrbourbeau replied:
Gotcha. FWIW it reads a little weird to me, so I have a slight preference for switching to `type` instead of `type_` (similar to what we have in the [`FutureState` class](https://github.com/dask/distributed/blob/946c6a46abb04615051537f1bf39b81163aa91f0/distributed/client.py#L432))

gforsyth commented on distributed/worker.py:
```diff
@@ -1508,35 +1506,31 @@ def transition_dep_waiting_flight(self, dep, worker=None):
-                if self.task_state[key] == "waiting":
+            if not ts.who_has:
+                if ts.key not in self._missing_dep_flight:
+                    self.loop.add_callback(self.handle_missing_dep, ts)
```
Nice catch!  Thanks!

gforsyth commented on distributed/worker.py:
```diff
@@ -1566,55 +1566,26 @@ def transition_dep_flight_memory(self, dep, value=None):
+                assert ts.key not in self.ready
 
-            self.waiting_for_data.pop(key, None)
+            ts.waiting_for_data = set()
```
Definitely.  I thought I'd changed all of these but missed this one.  Thanks!

gforsyth commented on distributed/worker.py:
```diff
@@ -1792,31 +1770,32 @@ def ensure_communicating(self):
                     self.loop.add_callback(self.handle_missing_dep, *missing_deps2)
 
-                    deps = [dep for dep in deps if dep not in missing_deps]
+                    deps = [dep for dep in deps if dep.key not in missing_deps]
```
Yep, tested and confirmed this.

jrbourbeau reviewed (commented):
Thanks for the updates @gforsyth! It looks like there are a few remaining `executing` -> `executing_count` updates, but other this looks good to go

mrocklin commented:
Passed tests here are encouraging.  I'm going to restart Github actions CI to see if this is consistent.

jrbourbeau reviewed (approved):
Thanks @gforsyth! I'm +1 on merging the changes here. I'll leave this open for a bit to allow for last minute feedback, but then I plan on merging
